File: qryoklama_flutter/lib/service/location_service.dart
import 'package:geolocator/geolocator.dart';

class LocationService {
  Future<Position?> getBestLocation() async {
    // izin
    var perm = await Geolocator.checkPermission();
    if (perm == LocationPermission.denied) {
      perm = await Geolocator.requestPermission();
    }
    if (perm == LocationPermission.denied ||
        perm == LocationPermission.deniedForever) {
      return null;
    }

    // servis açık mı
    final enabled = await Geolocator.isLocationServiceEnabled();
    if (!enabled) return null;

    // güncel dene
    try {
      final pos = await Geolocator.getCurrentPosition(
        desiredAccuracy: LocationAccuracy.high,
        timeLimit: const Duration(seconds: 8),
      );
      if (isUsable(pos)) return pos;
    } catch (_) {}

    // last known fallback
    try {
      final last = await Geolocator.getLastKnownPosition();
      if (last != null && isUsable(last)) return last;
    } catch (_) {}

    return null;
  }

  /// Android isLocationUsable ile aynı:
  /// - 0,0 değil
  /// - 2dk’dan eski değil
  /// - accuracy 100m’den kötü değil
  bool isUsable(Position p) {
    if (p.latitude == 0.0 && p.longitude == 0.0) return false;

    final ts = p.timestamp;
    if (ts != null) {
      final age = DateTime.now().difference(ts);
      if (age > const Duration(minutes: 2)) return false;
    }

    if (p.accuracy > 100) return false;

    return true;
  }
}

File: qr_yoklama_react_native/src/service/locationService.ts
import * as Location from "expo-location";

export type SimpleLocation = { lat: number; lng: number };

export const LocationService = {
    async getBestLocation(): Promise<SimpleLocation | null> {
        const perm = await Location.requestForegroundPermissionsAsync();
        if (perm.status !== "granted") return null;

        const enabled = await Location.hasServicesEnabledAsync();
        if (!enabled) return null;

        try {
            const cur = await Location.getCurrentPositionAsync({
                accuracy: Location.Accuracy.High,
            });
            if (this.isUsable(cur)) {
                return { lat: cur.coords.latitude, lng: cur.coords.longitude };
            }
        } catch (_) {}

        try {
            const last = await Location.getLastKnownPositionAsync();
            if (last && this.isUsable(last)) {
                return { lat: last.coords.latitude, lng: last.coords.longitude };
            }
        } catch (_) {}

        return null;
    },

    isUsable(pos: Location.LocationObject): boolean {
        const { latitude, longitude, accuracy } = pos.coords;
        if (latitude === 0 && longitude === 0) return false;

        if (pos.timestamp) {
            const ageMs = Date.now() - pos.timestamp;
            if (ageMs > 120_000) return false;
        }

        if (typeof accuracy === "number" && accuracy > 100) return false;

        return true;
    },
};
